Paper: Prabhu Ramachandran: FOSSEE paper

prabhuramachandran opened the pull request:
Status: the paper is reasonably complete in terms of the content but requires a few revisions.

kyleniemeyer commented:
Hi there, just FYI—I've been asked to review this paper

prabhuramachandran commented:
@kyleniemeyer -- Hi!  Thanks for taking the time to review this!

suchow commented:
Here's some advice from Proselint (http://github.com/amperser/Proselint), a Python-based linter for prose:

```
184:69: consistency.spacing Inconsistent spacing after period (1 vs. 2 spaces). Found 12 times elsewhere.
291:28: consistency.spelling Inconsistent spelling of 'focused' (vs. 'focussed').
593:1: redundancy.garner Redundancy. Use 'plans' instead of 'Future plans'.
650:1: redundancy.garner Redundancy. Use 'plans' instead of 'Future plans'.
184:18: weasel_words.very Substitute 'damn' every time you're inclined to write 'very;' your editor will delete it and the writing will be just as it should be. Found 11 times elsewhere.
```

sarostru commented:
FOSSEE is really doing some amazing work!  Creating open source replacements for technical software and creating course material is an immense undertaking.

I was happy to learn about the various initiatives and events that have been done over the last 5 or 6 years.

I do however think you may want to reconsider how the paper is structured.  I think it is too broad in scope and currently it reads like a magazine article, listing all of the different achievements and projects that have been done.  Outside of them having been done by people under the umbrella of FOSSEE there isn't always much relation between them and they don't build up a single idea.  It is fine to have a paragraph or section that is an overview of FOSSEE and a list of its various achievements, but not the whole paper.

The first question to ask yourself, is what will a reader learn from this paper?  It should be a crisp one or two sentence proposition that can draw in the reader, my personal thought is that a tighter focus for the paper might be:

- How to teach a class to program using fully open source tools?

Choosing a core question like this focuses your ability to write the paper, since now you have a specific audience; people who teach classes, which is possibly almost everyone at scipy.  This lets you weigh whether or not any given piece of information belongs in the paper based on its perceived value to that audience.  Additionally, we can now directly speak of the value to a reader and to science, since your paper (and FOSSEE) is helping educators get more students programming (i.e. clear scientific value).

I think you could focus more on the flow of how the different tools you provide fit into a framework of a course being taught.  You get into some of these connections with Yaksh and I think you could do the same for the other tools and present them as a cohesive set of tools for running a class.  

For instance spoken tutorials and textbook companions are both great ideas, but as an instructor how would they be incorporated into a course?  Are they extras for the keeners or do they help students who are struggling?  Do they replace live lectures?  How do students access them, workstations at the University or can they view them at home?

In the Yaksh section there seems to be some scope creep, I'm not sure if a step by step tutorial is needed, it feels out of place with the rest of the content which is very high level.  Similarly the details of how Yaksh launches servers, what protocol it uses, and what user it runs as are not very interesting or important.  Very valuable though are showing some example questions and usage and the user experience story of how effective Yaksh can be and how to use it effectively.

To complement the focus of this being a more scientific paper, you should reference related material more, especially in the beginning.  So who else is doing open source course creation?  In what ways does this approach differ?  Do you collaborate?

I think you can safely drop the listing of previous scipy india conferences doesn't teach the reader anything.

Some general notes throughout:

- needs to be spell and grammar checked
- consistent perspective:
  - it switches between passive voice, we, and I at different points in the paper.  You should stick with either we or passive voice everywhere.  If you want to embed a personal story, you should present it as a quote.
 - acronyms need to be introduced first, e.g. EDA, GAE, CBSE

I hope this is helpful, looking forward to seeing a revised version!

kyleniemeyer commented:
I agree with a lot of what @sarostru said—FOSSEE sounds like a great initiative, and has clearly done great work. I will second @sarostru's comments that it reads more like a magazine article or blog post in a lot of places than a technical article, with many persuasive elements (e.g., "submitted textbook companions are very well made", "online tests fun and very helpful").

I do think many people at SciPy could be interested in hearing about these efforts, but right now the paper doesn't have a clear focus on a topic, particularly one relevant to SciPy. In fact, there really isn't a connection made to scientific computing at all. (In fact, while reading the paper, I thought it might be more appropriate for a talk at PyCon, or in particular the PyCon Education Summit.) The paper starts describing various components and achievements of FOSSEE, but don't really get into details other than with Yaksh. I do have a few concerns about some overlap with the 2014 SciPy paper (reference `[kmm14]`), which is cited as providing more detail in a few places. Then, at the end, there's a plug for SciPy India that doesn't seem to fit with the rest of the paper.

Beyond that, I encourage you to follow @sarostru's suggestion of focusing on one aspect of what's in the current version of the paper, rather than a very broad description of FOSSEE and related activities. Keep in mind the focus of SciPy, which is on scientific computation.

A few general comments: 
 * In the abstract, could replace "has for its goal" just "its goal is"
 * In introduction, "the goal to try help minimize" -> "the goal of minimizing" or "the goal of helping to minimize"
 * In introduction, "sterling" is a strange word choice
 * In introduction, the initiatives that worked vs. those that didn't might be of interest
 * What is the "novice check" that spoken-tutorial submissions undergo?
 * "One oft quoted problem with FOSS tools is the lack of documentation"—do you have anything to back this up? This is a fairly general statement to make, and considering the Python/SciPy audience you would likely get some pushback here. However, there's no question that effective documentation is important.
 * What is the signal-to-noise ratio of submissions for the textbook companions? Meaning, how many are accepted vs. submitted?
 * The strange indent at the beginning of the numbered list in the textbook companions section can likely be fixed by removing the space preceding the numbers (I had the same problem in my paper, and this fixed it.)
 * In item 1 of that list, "textbooks that has not" -> "textbooks that have not"
 * Bottom of page 3: "fairtly" -> "fairly"
 * How was it determined that some of the submitted textbook companions are very well made? This seems pretty subjective.
 * How different is the material being taught in the SDES course from the Software Carpentry material?

kyleniemeyer commented:
I should emphasize @sarostru's point about referencing more related work near the beginning would strengthen the paper, by allowing you to place your work in the context of what other are doing or have done.

taldcroft commented:
I will respectfully disagree with the first two reviewers that the scope of paper should be narrowed.  I found each of the different sections (all under the umbrella of FOSSEE) to be quite fascinating and would be disappointed to see any of them cut from the final paper [1].  I have some involvement in educating students in Python and scientific computing, and I found all of the presented topics and the lessons-learned to be relevant and informative.  The foundation for any student doing scientific Python computing is the set of general Python and other "software carpentry" skills discussed here in different contexts, so for me this paper is well suited in focus for the SciPy conference proceedings.

Like the others am impressed at the scope of work and the execution.

The textbook companions are an interesting concept, but I also wonder how this plays out in practice in terms of providing benefit to the educational process.  The paper mentions that the concept will be made public soon and you will analyze usage of the notebooks, but do you have thoughts about what to tell instructors in terms of using them?  Also, are there potential copyright/legal issues associated with coding up textbook examples?  In the U.S. at least, publishers are in the habit of issuing new textbook revisions quite frequently and so there might be a problem with the companions getting out of date.

[1] - Except I agree that the section on SciPy India seems out of place, and the section on "Installation and running a demo" is incommensurate with the higher-level description in the rest of the paper and could be cut.

davclark commented on papers/prabhu_ramachandran_fossee/prabhu_ramachandran_fossee.rst:
```diff
@@ -0,0 +1,710 @@
+   and users can also post questions on a forum. :label:`fig:py-st`
+
+These spoken tutorials can be accessed by anyone and can also be downloaded
+into a self-contained CD by users.  Around 40000 users have gone over this
```
How did you measure this?

davclark commented on papers/prabhu_ramachandran_fossee/prabhu_ramachandran_fossee.rst:
```diff
@@ -0,0 +1,710 @@
+effective.  Instant feedback makes the process fun for the student.  The
+ability to submit multiple times gives them comfort in that they know that
+they can gradually fix their code.  This makes students less anxious.  They
+also immediately know that their answer is correct if they get it right.  This
```
I'd push back a bit on this - fully test-driven development can lead to randomly changing things until they work on the test cases. Understanding and "correctness" might still not be there. This is a very useful approach (I think), but best not to oversell it!

davclark commented:
First - apologies for the late review. I have been wrapping up a job, driving across country and now starting at Kennedy Krieger Institute. Therefore, I'll try to be brief, and will attempt to avoid duplicating previous comments.

## Summary

The paper provides a nice summary for some FOSSEE activities, of which I am a bit of a fanboy, so note that I am a bit biased here. Given that stipulation, I'd be sad to see this paper missed in the proceedings, and would be comfortable publishing it in it's current state *if necessary*. However, I think it could be substantially improved.

## Detailed notes

I resonate with both sides of the discussion above on focus of the paper. Certainly, the SciPy India section could be reduced to a link where interested parties could find out more about the history of the conference.

Regarding the rest, I agree that it comes across as unfocused. My suggestion (different again from the above!) is to spend some time on abstract and introduction and clarify a few sentences describing what the reader is intended to take home. Having done that, I think you could easily edit the paper into a more focused state (and we reviewers would be more able to judge how well you achieve this goal).

Given the crowd, I suspect the audience will be most interested in the tools developed to support teaching (along the lines that @sarostru suggests).

Coming from a data science education initiative at a major US university, I certify that these accomplishments are nothing short of incredible. *I* would like to see some more elaboration of the process of successes and failures that led to the current set of principles. Some of them resonate with what I've found ([shameless plug on my own efforts](http://datascience.codata.org/articles/10.5334/dsj-2016-005/) - though I'm actually not proud of this paper and am in the process of managing a corrigendum process! I've also done [a more recent iteration](http://hackingmeasurement.berkeley.edu/projects.html)) - scoping projects is important, and letting students develop their own projects is dangerous.

My guess is that (as in most cases), your data collection process was insufficient to have a good social science / education research exposition. But if you *do* have good data, you should share it!

prabhuramachandran commented:
Thank you @sarostru, @kyleniemeyer, @taldcroft, and @davclark for your detailed comments.  I am sorry I have nor responded thus far. I was traveling and was working on the review comments on the other submission which is now done!   That paper is much more technical and has gone smoothly.  I must admit that I am not exactly sure how best to write up the current paper so what I had put together was really more like a magazine article.   I will rethink this a bit and see what I can do to make it more focussed although I admit it is a little hard for me.  Your suggestions should give me a few ideas though. I am hoping to have some changes based on your advice.

prabhuramachandran commented:
Hi  @sarostru, @kyleniemeyer, @taldcroft, and @davclark I went over your comments and completely agree that a central focus is lacking.  The language/stylistic issues are easier to fix.  I am going to modify the text to be more centered around spreading the use of Python in collegiate education and focus the discussion on the following:

 1. Teaching at a larger scale with a large number of students and our experiences (brings to focus the spoken-tutorials).
 2. The SDES course as a means to teach students better computing.
 3. Yaksh as a means to effectively test students at scale
 4. The textbook companions to allow a large number of user contributed documents.
 5. SciPy India as a forum for people to meet and learn.

I guess this will provide a good story around the loosely organized pieces I have slapped together thus far.  Doing this will not take too much effort I imagine. I'll try and submit a revision by today.  In the meanwhile if you have any suggestions beyond this I will be happy to incorporate these.  Thanks!

sarostru commented:
Thanks @prabhuramachandran, that sounds like a good approach to me, looking forward to a second set of changes.

prabhuramachandran commented:
Thanks, I am traveling again today and will make the changes to the remaining two sections and submit it tonight.  This should hopefully fix the embarrassing spelling and grammar errors from before and also reflect the reorganization as I have mentioned above.  I do not have the resources to dig up additional references at this point though so I am not sure I will be able to add a whole lot of reference material now as I am traveling.

prabhuramachandran commented:
Hi all, I have refactored the paper quite a bit as per what I have written above.  Here are some responses to specific points:

> needs to be spell and grammar checked

Done.  Apologies for this, I didn't use flyspell-mode properly on this paper.

> consistent perspective:
> it switches between passive voice, we, and I at different points in the paper. You should stick with either > we or passive voice everywhere. If you want to embed a personal story, you should present it as a quote.

Done.

> acronyms need to be introduced first, e.g. EDA, GAE, CBSE

Done.

> In the abstract, could replace "has for its goal" just "its goal is"

Done.

> In introduction, "the goal to try help minimize" -> "the goal of minimizing" or "the goal of helping to minimize"

Done.

> In introduction, "sterling" is a strange word choice

Fixed.

> In introduction, the initiatives that worked vs. those that didn't might be of interest

I haven't really added this in the introduction but where relevant I have added a few points of what did and what did not work.

>What is the "novice check" that spoken-tutorial submissions undergo?

I have added a bit of explanation.

> "One oft quoted problem with FOSS tools is the lack of documentation"—do you have anything to back > this up? This is a fairly general statement to make, and considering the Python/SciPy audience you >would likely get some pushback here. However, there's no question that effective documentation is >important.

This has been completely rewritten, thanks.

> What is the signal-to-noise ratio of submissions for the textbook companions? Meaning, how many are accepted vs. submitted?

I have added some approximate statistics for this.

> The strange indent at the beginning of the numbered list in the textbook companions section can likely be fixed by removing the space preceding the numbers (I had the same problem in my paper, and this fixed it.)

Fixed.

> In item 1 of that list, "textbooks that has not" -> "textbooks that have not"
> Bottom of page 3: "fairtly" -> "fairly"

Fixed. Sorry about the embarrassing errors.

> How was it determined that some of the submitted textbook companions are very well made? This seems pretty subjective.

Some examples are just simple translations and some take extra effort to make a nice plot etc.  These are relatively easy to see.  This is definitely subjective and that is why those are called "editor's picks".

> How different is the material being taught in the SDES course from the Software Carpentry material?

I have explained this a bit.  The content is really not too different except that this is a course that students can credit.

Finally, I have tried to address all comments I have seen on the manuscript so far and have cut out some of the unnecessary material on the scipy India conference but have retained most of the real content since it will be useful to some readers.  I have tried to weave a story around all of these focussed on effective teaching of Python and spreading the adoption of Python.  Thanks for all your comments.  I will not be able to make more changes until this weekend as I am traveling.

sarostru commented:
Due to the complication of the proceedings deadline lining up with everyone's travel schedule to attend the conference we've decided to allow authors to spend an additional 2 days, that is until July 12th to complete any major changes to the submission.  This may impact your enjoyment of the conference, so we hope not too many of you need to take the time.  For reviewers, we ask that you read over this final document before the conference is over and post your decision here in the comments on or before July 17th.

Thanks and have a great Scipy!
Your proceedings co-chairs Scott Rostrup (@sarostru) and Sebastian Benthall (@sbenthall)

sarostru commented:
I don't think my major concern has been addressed, namely that it reads like a magazine article and not like a scientific paper.  Trying to apply the suggested [review criteria](https://github.com/scipy-conference/scipy_proceedings/blob/master/review_criteria.md) to this is a bit like comparing apples and oranges, they are different things.

My feeling is that you are submitting to the wrong place, you could publish such an article in other places and it would be well received.  Perhaps a magazine, e.g. Linux Magazine, Full Circle, Linux For You or Open Source For You; or publishing online, on a blog or aggregator site.

For this reason, I recommend not accepting this submission to the scipy 2016 proceedings.

davclark commented:
To echo @sarostru, it seems this paper being in the proceedings is not a huge win. The talk can still happen, and the information might profitably be presented hypertextually on the FOSSEE site. This would allow for updating of relevant stats over time, etc.

taldcroft commented:
This is primarily an educational paper, and as in other science disciplines like astronomy, papers that focus on educational elements tend to have a somewhat different narrative style than purely technical papers.  I think it would be a disservice the educational and SciPy community to not publish this paper in the SciPy proceedings because it does not sound "scientific" enough.

We can compare this paper to other previously published SciPy conference proceedings papers.  Fortunately SciPy 2014 was focused on education and there are many examples.  For example, I think that [Scientific Computing with SciPy for Undergraduate Physics Majors](http://conference.scipy.org/proceedings/scipy2014/baxter.html) shares a similar style in that it provides a narrative description of a particular educational process.  I would say that the FOSSEE paper has significantly more depth and informative technical content, including links to original software packages and a demo material (which was shortened at the request of reviewers).  Another paper from 2014 [Teaching numerical methods with IPython notebooks and inquiry-based learning](http://conference.scipy.org/proceedings/scipy2014/ketcheson.html) seems relatively comparable to the FOSSEE paper.

I agree that the current FOSSEE paper is a bit of hodge-podge and does not have just one single theme, but the point is to report on several related efforts under the umbrella of FOSSEE. 

I accept this paper for merging.

prabhuramachandran commented:
Thanks all for the comments.  I am fine with whatever decision you folks choose to take.  I will defer to the proceedings chairs to decide as they wish.  I am grateful to all of you to have put in so much time into this effort of reviewing this and am sorry if it is not good enough to be part of the proceedings and am glad you all care about the quality of the proceedings!

prabhuramachandran commented:
@taldcroft -- I agree with you in that it is not easy for me to write this in the same narrative as a standard engineering/science paper as the approach has not been that "scientific" and I agree the paper is not focussed.  That said the proceedings chairs should make the decision that is correct in their minds and that is perfectly OK with me.

sarostru commented:
That is a really good point about the education focussed papers, @taldcroft, it makes me reconsider my decision.  I think for future reviewers and proceedings organizers we should update the [review criteria](https://github.com/scipy-conference/scipy_proceedings/blob/master/review_criteria.md) to provide guidelines for papers with more of an educational focus, @katyhuff @stefanv @sbenthall.

@prabhuramachandran you have made a lot of the changes we wanted and the contributions of FOSSEE to the python community seem amazing, so yes I accept this paper for merge.

kyleniemeyer commented:
Yes, I agree that @prabhuramachandran made most of the changes we recommended, and therefore we shouldn't change the goalposts after the fact—I'm comfortable accepting at this point.

I do agree that the review/submission guidelines could be made more clear on contributions of this nature (whether positively or negatively).

prabhuramachandran commented:
Hi everyone, as I said, I am perfectly OK with whatever you folks decide.  Thank you all for reconsidering and I am grateful for your openness.

davclark commented:
Just to be clear, I'd already said:

> ...I am a bit of a fanboy, so note that I am a bit biased here. Given that stipulation, I'd be sad to see this paper missed in the proceedings, and would be comfortable publishing it in it's current state *if necessary*. However, I think it could be substantially improved.

Upshot, I still approve a merge.

After some thinking, I became less sure that it's best if this is in the proceedings or simply on the web as a resource, but it seems like @prabhuramachandran would still like to get it published if that's OK with reviewers. I am totally fine with that.
